File: tennis-game/tennis-game-kotlin-arrow/src/test/src/GamePointPlayTest.kt
import GameInteraction.displayableGameScore
import GamePlay.trackScorePointForPlayer
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test

class GamePointPlayTest {

    private lateinit var game: Game

    @Before
    fun setUp() {
        game = Game()
    }

    @Test
    fun `player 1 score fifteen`() {
        val updatedGame = trackScorePointForPlayer(Player1, game)
        val gameScore = displayableGameScore(updatedGame)
        assertThat(gameScore, `is`(equalTo("Player 1 Fifteen - Player 2 Love")))
    }

    @Test
    fun `player 1 score Thirty`() {
        val gameAfterFirstPlay = trackScorePointForPlayer(Player1, game)
        val gameAfterSecondPlay = trackScorePointForPlayer(Player1, gameAfterFirstPlay)

        val gameScore = displayableGameScore(gameAfterSecondPlay)

        assertThat(gameScore, `is`(equalTo("Player 1 Thirty - Player 2 Love")))
    }

    @Test
    fun `player 1 score Forty`() {
        val updatedGame =
                trackScorePointForPlayer(Player1, game)
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 Forty - Player 2 Love")))
    }

    @Test
    fun `player 2 score Fifteen`() {
        val gameAfterFirstPlay = trackScorePointForPlayer(Player2, game)

        val gameScore = displayableGameScore(gameAfterFirstPlay)

        assertThat(gameScore, `is`(equalTo("Player 1 Love - Player 2 Fifteen")))
    }

    @Test
    fun `player 2 score Thirty`() {
        val updatedGame =
                trackScorePointForPlayer(Player2, game)
                        .let { g -> trackScorePointForPlayer(Player2, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 Love - Player 2 Thirty")))
    }

    @Test
    fun `player 2 score Forty`() {
        val updatedGame =
                trackScorePointForPlayer(Player2, game)
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 Love - Player 2 Forty")))
    }

    @Test
    fun Deuce() {
        val updatedGame =
                trackScorePointForPlayer(Player1, game)
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Deuce")))
    }

    @Test
    fun `player 1 wins`() {
        val updatedGame =
                trackScorePointForPlayer(Player1, game)
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 wins")))
    }

    @Test
    fun `player 2 wins`() {
        val updatedGame =
                trackScorePointForPlayer(Player2, game)
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 2 wins")))
    }

    @Test
    fun `player 1 advantage`() {
        val updatedGame =
                trackScorePointForPlayer(Player1, game)
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 Advantage - Player 2 Forty")))
    }

    @Test
    fun `player 2 advantage`() {
        val updatedGame =
                trackScorePointForPlayer(Player1, game)
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player1, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }
                        .let { g -> trackScorePointForPlayer(Player2, g) }

        val gameScore = displayableGameScore(updatedGame)

        assertThat(gameScore, `is`(equalTo("Player 1 Forty - Player 2 Advantage")))
    }
}

File: tennis-game/typescript-fpts-tennis-game/__test__/GamePointPlay.test.ts
import {trackScorePointForPlayer} from "../src/GamePlay";
import {createGame, Game, player1, player2} from "../src/Data";
import {displayableGameScore} from "../src/GameInteraction";
import {pipe} from "fp-ts/pipeable";

describe('GamePointPlay', () => {
    const trackScorePlayer1 = (game: Game) => trackScorePointForPlayer(player1(), game)
    const trackScorePlayer2 = (game: Game) => trackScorePointForPlayer(player2(), game)

    describe('player 1',  () => {
        it('score fifteen', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                displayableGameScore
            )

            expect(gameScore).toEqual("Player 1 Fifteen - Player 2 Love")
        });

        it('score thirty', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                trackScorePlayer1,
                displayableGameScore
            )

            expect(gameScore).toEqual("Player 1 Thirty - Player 2 Love")
        });

        it('score forty', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                trackScorePlayer1,
                trackScorePlayer1,
                displayableGameScore
            )
            expect(gameScore).toEqual("Player 1 Forty - Player 2 Love")
        });

        it('wins', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer1,
                displayableGameScore
            )
            expect(gameScore).toEqual("Player 1 wins")
        });

        it('advantage', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer1,
                displayableGameScore
            )
            expect(gameScore).toEqual("Player 1 Advantage - Player 2 Forty")
        });
    })

    describe('player 2',  () => {
        it('score fifteen', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer2,
                displayableGameScore
            )

            expect(gameScore).toEqual("Player 1 Love - Player 2 Fifteen")
        });

        it('score thirty', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer2,
                trackScorePlayer2,
                displayableGameScore
            )

            expect(gameScore).toEqual("Player 1 Love - Player 2 Thirty")
        });

        it('score forty', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer2,
                trackScorePlayer2,
                trackScorePlayer2,
                displayableGameScore
            )

            expect(gameScore).toEqual("Player 1 Love - Player 2 Forty")
        });


        it('wins', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer2,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer2,
                displayableGameScore
            )
            expect(gameScore).toEqual("Player 2 wins")
        });

        it('advantage', () => {
            const gameScore = pipe(
                createGame(),
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer1,
                trackScorePlayer2,
                trackScorePlayer2,
                displayableGameScore
            )
            expect(gameScore).toEqual("Player 1 Forty - Player 2 Advantage")
        });
    })

    it('Deuce', () => {
        const gameScore = pipe(
            createGame(),
            trackScorePlayer1,
            trackScorePlayer1,
            trackScorePlayer1,
            trackScorePlayer2,
            trackScorePlayer2,
            trackScorePlayer2,
            trackScorePlayer1,
            trackScorePlayer2,
            displayableGameScore
        );

        expect(gameScore).toEqual("Deuce")
    });
});
